rat/codegen_backends.py

```diff
@@ -1,357 +1,269 @@
-                    self.at_rhs_of_scan = False
-                    self.expression_string += "\n"
-                    self.expression_string += f"    return {next_carry_tuple_string}, next_value\n\n"
-
-                    self.expression_string += "_, "
-                    self.generate(lhs)
-                    self.expression_string += f" = jax.lax.scan({scan_function_name}, {zero_tuple_string}, jax.numpy.arange({lhs_size}))"
-                    self.expression_string += "\n"
+                    index_lookup = ""
+
+                return f"parameters['{node.name}']{index_lookup}"
 
+    def walk_Literal(self, node: ast.Literal):
+        return f"{node.value}"
+
+
+class CodeExecutor(NodeWalker):
+    arguments: Dict[ast.Variable, Any]
+    parameters: Dict[str, Any]
+    left_side_of_sampling: ast.ModelBase
+
+    def __init__(self, arguments: Dict[ast.Variable, Any] = None, parameters: Dict[str, Any] = None):
+        self.arguments = arguments
+        self.parameters = parameters
+        self.left_side_of_sampling = None
+
+    def walk_Statement(self, node: ast.Statement):
+        if node.op == "~":
+            self.left_side_of_sampling = node.left
+            return_value = self.walk(node.right)
+            self.left_side_of_sampling = None
+            return return_value
+        else:
+            raise CompileError(f"{node.op} operator not supported in BaseCodeGenerator", node)
+
+    def walk_Logical(self, node: ast.Logical):
+        left = self.walk(node.left)
+        right = self.walk(node.right)
+        return eval(f"(left {node.op} right)")
+
+    def walk_Binary(self, node: ast.Binary):
+        left = self.walk(node.left)
+        right = self.walk(node.right)
+        return eval(f"left {node.op} right")
+
+    def walk_IfElse(self, node: ast.IfElse):
+        predicate = self.arguments[node.predicate]
+        return jax.lax.cond(predicate, lambda: self.walk(node.left), lambda: self.walk(node.right))
+
+    def walk_FunctionCall(self, node: ast.FunctionCall):
+        arglist = []
+
+        if self.left_side_of_sampling:
+            arglist += [self.walk(self.left_side_of_sampling)]
+
+        if node.arglist:
+            arglist += [self.walk(arg) for arg in node.arglist]
+
+        return eval(node.name)(*arglist)
+
+    def walk_Variable(self, node: ast.Variable):
+        if node in self.arguments:
+            if node.name in self.parameters:
+                trace = self.arguments[node]
+                return self.parameters[node.name][trace]
+            else:
+                return self.arguments[node]
+        else:
+            return self.parameters[node.name]
+
+    def walk_Literal(self, node: ast.Literal):
+        return node.value
+
+
+class TraceExecutor(NodeWalker):
+    """
+    As long as shunt is true, values of expressions aren't returned but
+    instead shunted off into tuples
+    """
+
+    tracers: Dict[str, Tracer]
+    leaves: Dict[str, Any]
+    shunt: ContextStack[bool]
```

Comment:
I don't understand how the exact behavior of `shunt` is related with respect to `first_level_values`. I see that shunt is used in variable subscripts to indicate that values should be evaluated and returned instead of being traced(recorded), but how is first_level_values related?

Reply:
It's awkward that so much happens here -- I think it wouldn't be bad to split this in two except then I'd be repeating myself for all but a couple bits of logic :/.